janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,146 @@
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() { return INSTANCE; }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        TraversalHelper.getStepsOfClass(HasStep.class, traversal)
+            .forEach(this::optimizeStep);
+    }
+
+    /**
+     * Determines whether this HasStep can be answered by a unique index and thus, returns either 0 or 1 match
+     */
+    @Override
+    protected boolean isValidStep(HasStep<?> step) {
+        StandardJanusGraphTx tx;
+
+        try {
+            tx = (StandardJanusGraphTx) JanusGraphTraversalUtil.getTx(step.getTraversal());
+        } catch (IllegalArgumentException iaex) {
+            // The traversal needs to be bound to a graph in order to get information about the indexes
+            return false;
+        }
+
+        List<String> givenKeys = step.getHasContainers().stream()
+            .map(HasContainer::getKey).collect(Collectors.toList());
+
+        List<PredicateCondition<String, JanusGraphElement>> constraints = step.getHasContainers().stream()
+            .filter(hc -> hc.getBiPredicate() == Compare.eq)
+            .map(hc -> new PredicateCondition<>(hc.getKey(), JanusGraphPredicateUtils.convert(hc.getBiPredicate()), hc.getValue()))
+            .collect(Collectors.toList());
+        final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
+
+        // collect all matching indexes
+        final Set<IndexType> availableIndexes = new HashSet<>();
+        ConditionUtil.traversal(conditions, condition -> {
+            if (condition instanceof PredicateCondition) {
+                final RelationType type = ((PredicateCondition<RelationType,JanusGraphElement>) condition).getKey();
+                Preconditions.checkArgument(type != null && type.isPropertyKey());
+                Iterables.addAll(availableIndexes, ((InternalRelationType) type).getKeyIndexes());
+            }
+            return true;
+        });
+
+        // select only valid unique indexes
+        return availableIndexes.stream()
```

Comment:
Maybe you can put the selection logic into a new class that implements IndexSelectionStrategy interface